### PR TITLE
141581 jianpu

### PR DESCRIPTION
**Pull request by byan61**
This is initial implementation of Jianpu staff display and adding a linked Jianpu staff from a Standard staff.
The implementation is just a start and is by no means a complete one yet.
Currently display of octave dots is limited to one dot.
Supports for duration augmenting dots, grace notes, spanners, etc., are not there yet.
There are also some TODO items marked in the code.

### CODE REVIEW

**Comment by lasconic** [comment]
Hi, thank you for the PR. Two things:

1. I need you to sign the MuseScore CLA in order to pull your code into master https://musescore.org/en/cla

2. Your PR cannot be merged as it. It has several conflicts with master

```
libmscore/CMakeLists.txt
libmscore/barline.cpp
libmscore/element.h
libmscore/layout.cpp
libmscore/measure.cpp
libmscore/staff.cpp
libmscore/staff.h
libmscore/stafftype.cpp
libmscore/system.cpp
```

**Comment by byan61** [comment]
Done with #1.
For #2, I merged Jianpu changes with latest from master branch.
But, after the merge, Jianpu barlines and Jianpu beams do not work anymore.
I'm now working to fix these.

**Comment by byan61** [comment]
Merge conflicts have been resolved. It's ready for review.
Reading/writing Jianpu staff from/to the score file is not implemented yet.
So, if you test Jianpu staff display by adding a linked Jianpu staff, do not save your score file. Otherwise your score file may get corrupted.

**Comment by lasconic** [comment]
It seems your code cannot be compiled, see https://travis-ci.org/musescore/MuseScore/jobs/190151221#L1926

**Comment by byan61** [comment]
Sorry, some files were left out accidentally.
They have been pushed now.

**Comment by lasconic** [comment]
No problem. The CI is running. Next thing will be to squash all your commits together to not pollute the history of the repo. See http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html

**Comment by lasconic** [comment]
CI is still failing https://travis-ci.org/musescore/MuseScore/jobs/190154341#L1998

**Comment by Jojo-Schmitz** [comment]
I don't get why, but some mtests are failing, namely tst_midimapping, see https://travis-ci.org/musescore/MuseScore/jobs/190164161#L5105

File: libmscore/stafftype.cpp [hunk 1]
```diff
@@ -1384,6 +1386,7 @@ void StaffType::initStaffTypes()
          StaffType(StaffGroup::PERCUSSION, "perc1Line", QObject::tr("Perc. 1 line"),    1, -4,    1,   true,  true, false, true, false, true),
          StaffType(StaffGroup::PERCUSSION, "perc3Line", QObject::tr("Perc. 3 lines"),   3, 0,     2,   true,  true, false, true, false, true),
          StaffType(StaffGroup::PERCUSSION, "perc5Line", QObject::tr("Perc. 5 lines"),   5, 0,     1,   true,  true, false, true, false, true),
+         StaffType(StaffGroup::JIANPU,     "jianpu",    QObject::tr("Jianpu"),          0, 0,     1,  false,  true, false, true, false, false),
```
Comment (Jojo-Schmitz) [inline libmscore/stafftype.cpp]: I believe this is why the mtest fail, here you insert Jianpu

File: libmscore/stafftype.h [hunk 2]
```diff
@@ -155,6 +155,7 @@ enum class StaffTypes : signed char {
       TAB_6SIMPLE, TAB_6COMMON, TAB_6FULL,
             TAB_4SIMPLE, TAB_4COMMON, TAB_4FULL,
             TAB_UKULELE, TAB_BALALAJKA, TAB_ITALIAN, TAB_FRENCH,
+      JIANPU,
```
Comment (Jojo-Schmitz) [inline libmscore/stafftype.h]: whereas here you append jianpu

File: libmscore/stafftype.h [hunk 3]
```diff
@@ -243,6 +244,11 @@ class StaffType {
       static QList<TablatureDurationFont> _durationFonts;
       static std::vector<StaffType> _presets;
 
+      // Jianpu: configurable properties
+      // TODO: read them from configuration file.
+      QFont _jianpuNoteFont = QFont("Times New Roman", 11.0, QFont::Bold); // font used to draw Jianpu notes/rests.
```
Comment (Jojo-Schmitz) [inline libmscore/stafftype.h]: Better use fee serif or free sans

**Comment by byan61** [comment]
Thanks, Jojo.
I made the corrections and pushed the commit.

**Comment by Jojo-Schmitz** [comment]
Looks better ;-)
Now you'd just need to squash these commits (see @lasconic's hint above) , preferably with a commit title in the form of "fix #141581: implement Jianpu staff display", so the corresponding issue gets marked fixed automagically when this PR gets merged.

**Comment by Jojo-Schmitz** [comment]
Then again that squashing may wait until read/write has been implemented too

**Comment by byan61** [comment]
I'll be working on read/write part before squashing the commits.

**Comment by byan61** [comment]
After adding read/write and saving the score file with Jianpu linked staff, staves that are below Jianpu staff are not displayed at all when the file is opened.  Display of Jianpu staff and staves above it are working fine. The layout functions of elements from the staves in question are called but the draw functions do not get called. Verified with debugger that these elements are added to the BspTree by scanElements function. Any idea why draw function of these elements do not get called?

**Comment by Jojo-Schmitz** [comment]
A rebase is needed. And I guess this PR should get tagged as `Work in Progress`

**Comment by lasconic** [comment]
I quickly try the PR. I created a score for Flute, pressed i, and added a Jianpu linked staff. Then I entered some note in the standard flute staff. And I got this 

image

**Comment by byan61** [comment]
Note entry and editing for Jianpu has not been implemented yet, so this will not work.
Only display of Jianpu from existing Standard notation score file is implemented so far.
Try opening an existing score file and adding a linked Jianpu Staff to an existing Standard staff.

**Comment by byan61** [comment]
Arr... I just tried opening an existing score file and adding Jianpu staff and the display of Jianpu staff does not work anymore after merging with latest from the base branch.  I'll work to fix that.

**Comment by lasconic** [comment]
Sorry for that. It's true that the current master is a bit of a moving target right now. Feel free to ask for help if needed.

**Comment by byan61** [comment]
Sorry that I've been busy at work and at home and have not had much time to focus on this yet. Any help is welcome including pointing out what might be causing the Jianpu display issue.

**Comment by byan61** [comment]
Issue of Jianpu staff display has been fixed now.

**Comment by byan61** [comment]
@lasconic and @Jojo-Schmitz: Can you take a look and merge the PR if it's acceptable? Thanks.

**Comment by lasconic** [comment]
@byan61 sure, give us some time though, we are trying to release MuseScore 2.1 asap.

**Comment by Jojo-Schmitz** [comment]
These "merge remote-tracking branch" commist really shouldn't be there
You'd need to `git pull --rebase upstream/master` instead of `git merge ...`

**Comment by byan61** [comment]
Can you just be tolerant of these extra commits and continue with the code review?
It will take a lot of work (merge, build and test) unnecessarily just to remove these commits.
It is a waste of time.

**Comment by Jojo-Schmitz** [comment]
It is also a waste of time having to look through unneeded commits.

File: libmscore/jianpubeam.cpp [hunk 4]
```diff
@@ -18,6 +18,7 @@
 #include "chordrest.h"
 #include "chord.h"
 #include "rest.h"
+#include <assert.h>
```
Comment (Jojo-Schmitz) [inline libmscore/jianpubeam.cpp]: Should not be needed

File: libmscore/jianpubeam.cpp [hunk 5]
```diff
@@ -127,6 +240,9 @@ void JianpuBeam::layout()
             // Update y value for beams at next beam level.
             y += beamDistance;
             }
+      //qDebug("bbox x=%.0f y=%.0f w=%.0f h=%.0f", bbox().x(), bbox().y(), bbox().width(), bbox().height());
+      //assert(bbox().x() < 20000 && bbox().y() < 20000);
+      //assert(bbox().width() < 20000 && bbox().height() < 20000);
       }
```
Comment (Jojo-Schmitz) [inline libmscore/jianpubeam.cpp]: Better use Q_ASSERT() here.

**Comment by byan61** [comment]
Made changes to replace "assert" with "Q_ASSERT".  Please take a look.

File: libmscore/jianpunote.cpp [hunk 6]
```diff
@@ -58,7 +56,8 @@
 #include "hairpin.h"
 #include "textline.h"
 #include "durationtype.h"
-#include <assert.h>
+#include <QtGlobal>
```
Comment (Jojo-Schmitz) [inline libmscore/jianpunote.cpp]: Really needed?

**Comment by byan61** [comment]
The unnecessary header file is removed.

**Comment by byan61** [comment]
@lasconic and @Jojo-Schmitz: Can you take a look and merge the PR if it's acceptable so that I can move on to the next step? Thanks.

**Comment by Jojo-Schmitz** [comment]
Without you having squashed the commits?

**Comment by byan61** [comment]
I tried squashing the commits but there were a lot of conflicts, manual merges and could not finish it successfully since there were a lot of changes both in master and jianpu branch.  Is there any way from your admin level to do this automatically or easily?

**Comment by Jojo-Schmitz** [comment]
You could `git reset`to the last SAH before your changes, `git commit` your changes and then `git push -f`

**Comment by byan61** [comment]
Thanks for your instructions, Jojo.  And I have pushed the new commit.
But somehow CI build has following error:
Generating moc_abstractdialog.cpp
//qt5/bin/moc: /lib64/libc.so.6: version `GLIBC_2.14' not found (required by //qt5/bin/moc)
AUTOGEN: error: process for /MuseScore/build.release/mscore/moc_abstractdialog.cpp failed:

This error does not happen with my local build.
Please advise what might go wrong.

**Comment by Jojo-Schmitz** [comment]
forgot to add an UI file to the commit?

**Comment by Jojo-Schmitz** [comment]
Forgot to add an UI file to the commit? Like instrdialog.ui and instrwidget.ui?

**Comment by Jojo-Schmitz** [comment]
Also you didn't follow the proper format for the commit message, "fix #xxxxxx: some text"

**Comment by Jojo-Schmitz** [comment]
For sure stdfactory.cpp is missing

File: libmscore/CMakeLists.txt [hunk 7]
```diff
@@ -71,6 +71,8 @@ add_library (
       read206.cpp
       read300.cpp stafftypelist.cpp stafftypechange.cpp
       bracketItem.cpp
+      standardfactory.cpp jianpufactory.cpp jianpuchord.cpp jianpunote.cpp jianpurest.cpp
```
Comment (Jojo-Schmitz) [inline libmscore/CMakeLists.txt]: where is stdfactory.cpp?

File: libmscore/CMakeLists.txt [hunk 8]
```diff
@@ -71,6 +71,8 @@ add_library (
       read206.cpp
       read300.cpp stafftypelist.cpp stafftypechange.cpp
       bracketItem.cpp
+      standardfactory.cpp jianpufactory.cpp jianpuchord.cpp jianpunote.cpp jianpurest.cpp 
+      jianpuhook.cpp jianpubeam.cpp jianpunotedot.cpp
```
Comment (Jojo-Schmitz) [inline libmscore/CMakeLists.txt]: Seems you forgot to add all these new files to the commit

Comment (byan61) [inline libmscore/CMakeLists.txt]: Thanks, Jojo.  It was my dumb mistake to have forgotten to add new files.  It is fixed now in the new push. Please take a look.

**Comment by Jojo-Schmitz** [comment]
There are a couple compiler warnings that your coded introduced, please check and fix those.

**Comment by byan61** [comment]
I fixed the compiler warnings and pushed a new commit.  Please take a look.

**Comment by Jojo-Schmitz** [comment]
Still a few left:
/home/travis/build/musescore/MuseScore/libmscore/jianpunotedot.cpp:28:59: warning: unused parameter â€˜linkâ€™ [-Wunused-parameter]
 JianpuNoteDot::JianpuNoteDot(const NoteDot& noteDot, bool link)
                                                           ^
/home/travis/build/musescore/MuseScore/libmscore/jianpunotedot.cpp:33:65: warning: unused parameter â€˜linkâ€™ [-Wunused-parameter]
 JianpuNoteDot::JianpuNoteDot(const JianpuNoteDot& noteDot, bool link)

**Comment by byan61** [comment]
I have fixed the remaining compiler warnings and pushed a new commit, hopefully the last one for this PR. Please take a look.

**Comment by byan61** [comment]
@lasconic and @Jojo-Schmitz: Can you take a look and merge the PR if it's acceptable so that I can move on to the next step? Thanks.

**Comment by Jojo-Schmitz** [comment]
@byan61: code looks OK to me, I haven't actually tested any of it though. And I can't merge.

**Comment by byan61** [comment]
@lasconic and @Jojo-Schmitz: It's been a month now.  Any chance for testing and to get it merged?

**Comment by sirsoweird** [comment]
@lasconic @Jojo-Schmitz @byan61 I'd love to help test this, however I had a hard time getting a build going for either Windows or Linux.  Could anyone put something together for Windows so a few of us could test and give feedback? Sorry I'm not an expert programmer, but I'll help where I can.

**Comment by sirsoweird** [comment]
I finally got my build up and running.  Just giving it a few quick tests.
capture

A few thoughts: in the top line, the Jianpu underline went down when the notes went above B
Standard Jianpu would have the under-dots below the under-lines.  
Measure 7 shows a dotted half; the dot should not show for halves or longer notes.  The length of the note is only indicated by the number of dashes after the note.  For the dotted half, it should have two dashes after the note.

**Comment by lasconic** [comment]
Did you rebase on master and fix the conflicts? if yes, can you open a new PR ?

**Comment by sirsoweird** [comment]
No, I didn't rebase with master.  I'm still very interested in this project, however I'm not going to have much time to work on it for a while here.  Not sure where @byan61 is with this, he doesn't seem to be very active currently.  If we want to get back on track with this project, what would the steps be?

**Comment by lasconic** [comment]
Hi, 
First step would be to rebase to current master. Then fix the issues you mentioned, preferably with reference to an authoritative source. The problem with Jianpu is that there are several ways to notate it. If MuseScore includes one way, we will need a source to defend the implementation. Then the implementation needs to be clean up. The use of friend class makes it hard to read.

**Comment by dannysalim** [comment]
Hi, I am Danny Salim and am interested to continue this project.

**Comment by lasconic** [comment]
See PR #3614 for a rebased version.

**Comment by odhot** [comment]
> I finally got my build up and running. Just giving it a few quick tests.
> capture
> 
> A few thoughts: in the top line, the Jianpu underline went down when the notes went above B
> Standard Jianpu would have the under-dots below the under-lines.
> Measure 7 shows a dotted half; the dot should not show for halves or longer notes. The length of the note is only indicated by the number of dashes after the note. For the dotted half, it should have two dashes after the note.

hi, i have working on your version but for indonesian numbered notation .. where can i contact you sir ?

**Comment by odhot** [comment]
ss
here's the sample .. but our sharp and flat is different ..

**Comment by Jojo-Schmitz** [comment]
@odhot better go to #3614, as that is still open (but does need a rebase meanwhile)

**Comment by Buronan13** [comment]
@odhot boleh minta nomor wa atau ig atau nama fb kmu tidak?
Saya juga lagi cari cara membuat not angka di musescore tpi gak ketemu2?

**Comment by David-C0pperfield** [comment]
A year later...

**Comment by Jojo-Schmitz** [comment]
this is still closed and #3614 still open...

**Comment by Sunny090628** [comment]
[smusic.zip](https://github.com/musescore/MuseScore/files/4484398/smusic.zip)
Consider this for jianpu notation only, unlinked.

**Comment by odhot** [comment]
> [smusic.zip](https://github.com/musescore/MuseScore/files/4484398/smusic.zip)
> Consider this for jianpu notation only, unlinked.

different ... our numeric notation has different flag position 

Screen Shot 2020-04-16 at 09 11 22

**Comment by Thethnan** [comment]
> @lasconic @Jojo-Schmitz @byan61 I'd love to help test this, however I had a hard time getting a build going for either Windows or Linux. Could anyone put something together for Windows so a few of us could test and give feedback? Sorry I'm not an expert programmer, but I'll help where I can.

How to get this jianpu view note sheet.

**Comment by Thethnan** [comment]
I would like to get jianpu staff note to numbered note auto convert, please help me.

**Comment by Jojo-Schmitz** [comment]
this PR has been closed 2 1/2 years ago and work continued in #3614.
That in turn had been closed meanwhile too though, but if anything that'd be the one to continue on.

**Comment by oorfeo** [comment]
Last comment until now was Oct. 2020.
I'm Don Rechtman, a U.S. citizen currently residing in Shenzhen. More than one billion people here in China read, compose and perform using numbered Jianpu notation. If this project can be successfully completed, the Musescore community should receive a dramatic boost in size.
I'm available to beta and alpha any plugin work, and have resources to comment on suggestions for items for inclusion.

**Comment by Jojo-Schmitz** [comment]
See #3614, tht one is at least quite a bit newer

**Comment by oorfeo** [comment]
Thanks for the response!

 �

Yeah, that’s the one I was originally responding to. As a newbie to the discussion, it is a little difficult for me to understand exactly what was and is going on with the development of the plugin. 

 �

Also, having done programming, QA and technical writing in the past, I know the difficulties of implementation of seemingly simple ideas: Manager: “Move the OK button to the other side; can you do that this afternoon?” Programmer: “Actually, it’s going to take about two weeks to make it happen.”

 �

I’ve manually “translated” Western notation to Jianpu and back again for various music groups here. A lot of professional traditional Chinese singers are not trained in Western notation, and it also took me a fair amount of practice to master some of the more subtle aspects of Jianpu notation. Additionally, there are specialized characters and symbols for nuances on different instruments and even for different Ethnic and regional styles of singing. If the basics are able to be bullet-proofed (rhythm, pitch, key and tempo identification, multiple parts, lyrics), it should then be easier to customize for individual instruments and styles.

 �

Please do let me know if there’s any way I might assist with the development.

 �

Peace,

 �

Don

 �

Don “Orfeo” Rechtman �(唐瑞民)
53 Wutong Shan Shi Chang Keng Bei Cun, Luohu District, Shenzhen, Guangdong, China  518020
广东省深圳市罗湖区梧桐山市场坑背村53号
518020
Phone (China): �(+86) 135-2889-6144
Weixin (WeChat) name: � oorfeo
www.OrfeoMusic.org <http://www.orfeomusic.org/>  

“Delve into tradition with the utmost dedication and then break away from it with the utmost courage.”

Li Keran (1907-1989), a prominent figure in 20th-century Chinese art

 �



 �

**Pull request closed without merging**